src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -673,18 +685,27 @@ private Compilation CreateEmptyCompilation()
 
                         if (generatorDriver != null)
                         {
-                            generatorDriver = intermediateProject.action.TransformGeneratorDriver(generatorDriver);
+                            generatorDriver = action.TransformGeneratorDriver(generatorDriver);
                         }
 
                         // We have updated state, so store this new result; this allows us to drop the intermediate state we already processed
                         // even if we were to get cancelled at a later point.
                         intermediateProjects = intermediateProjects.RemoveAt(0);
 
-                        this.WriteState(CompilationTrackerState.Create(
-                            compilationWithoutGeneratedDocuments, state.GeneratorInfo.WithDriver(generatorDriver), compilationWithGeneratedDocuments, intermediateProjects));
+                        // As long as we have intermediate projects, we'll still keep creating InProgressStates.  But
+                        // once it becomes empty we'll produce an AllSyntaxTreesParsedState and we'll break the loop.
+                        var currentState = CompilationTrackerState.Create(
+                            compilationWithoutGeneratedDocuments, state.GeneratorInfo.WithDriver(generatorDriver), compilationWithGeneratedDocuments, intermediateProjects);
+                        this.WriteState(currentState);
+
+                        Contract.ThrowIfTrue(intermediateProjects.Count > 0 && currentState is not InProgressState);
+                        Contract.ThrowIfTrue(intermediateProjects.Count == 0 && currentState is not AllSyntaxTreesParsedState);
+
+                        resultState = currentState as AllSyntaxTreesParsedState;
                     }
```

Comment:
effectively this loop is always guaranteed to run.  and will do a series of `InProgress -> InProgress -> ... InProgress -> AllSyntaxTreesParsed` at which point it terminates.